Support multiple values in --component.

dblock opened the pull request:
Signed-off-by: dblock <dblock@amazon.com>

### Description
Added support for multiple values in `--component` in various build, test and sign workflows. 

I chose to support `--component x y` vs. `--component x --component y` to make it easier to specify from Jenkins. In code that's done with `nargs='+'` vs. `action='append'`, see https://stackoverflow.com/questions/15753701/how-can-i-pass-a-list-as-a-command-line-argument-with-argparse.

Together with https://github.com/opensearch-project/opensearch-build/pull/2100 this change allows on-demand rebuilding of a subset of components in a manifest when it contains a subsequent component that has a broken build. This is useful when wanting to increment versions and needing a new -SNAPSHOT build of OpenSearch or common-utils for dependencies that are already in a manifest. Avoids having to pull out a component from the manifest and then reverting that change every time.

### Issues Resolved

https://github.com/opensearch-project/opensearch-build/issues/2091

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

peterzhuamazon commented:
@dblock 
```
================================== FAILURES ===================================
________________ TestBuildManifest.test_select_two_are_unknown ________________

self = <tests.tests_manifests.test_build_manifest.TestBuildManifest testMethod=test_select_two_are_unknown>

    def test_select_two_are_unknown(self) -> None:
        path = os.path.join(self.data_path, "build", "opensearch-build-schema-version-1.2.yml")
        manifest = BuildManifest.from_path(path)
        with self.assertRaises(ValueError) as ctx:
            self.assertEqual(len(list(manifest.components.select(focus=["x", "y"]))), 0)
>       self.assertEqual(str(ctx.exception), "Unknown components=y,x.")
E       AssertionError: 'Unknown components=x,y.' != 'Unknown components=y,x.'
E       - Unknown components=x,y.
E       ?                     --
E       + Unknown components=y,x.
E       ?                    ++

tests\tests_manifests\test_build_manifest.py:113: AssertionError
```

dblock commented:
> @dblock
> 
> ```
> ================================== FAILURES ===================================
> ________________ TestBuildManifest.test_select_two_are_unknown ________________
> 
> self = <tests.tests_manifests.test_build_manifest.TestBuildManifest testMethod=test_select_two_are_unknown>
> 
>     def test_select_two_are_unknown(self) -> None:
>         path = os.path.join(self.data_path, "build", "opensearch-build-schema-version-1.2.yml")
>         manifest = BuildManifest.from_path(path)
>         with self.assertRaises(ValueError) as ctx:
>             self.assertEqual(len(list(manifest.components.select(focus=["x", "y"]))), 0)
> >       self.assertEqual(str(ctx.exception), "Unknown components=y,x.")
> E       AssertionError: 'Unknown components=x,y.' != 'Unknown components=y,x.'
> E       - Unknown components=x,y.
> E       ?                     --
> E       + Unknown components=y,x.
> E       ?                    ++
> 
> tests\tests_manifests\test_build_manifest.py:113: AssertionError
> ```

Fixed. Subtracting sets did not necessarily preserve order, so I fixed it with `invalid = [item for item in focus if item not in self]` vs. `invalid = focus - self.keys()`.

peterzhuamazon commented on src/build_workflow/build_args.py:
```diff
@@ -44,15 +48,34 @@ def __init__(self) -> None:
+            "-c",
+            "--component",
+            dest="components",
+            nargs='+',
```
This should be using `*` not `+`.

'+'. Just like '*', all command-line args present are gathered into a list. Additionally, an error message will be generated if there wasn’t at least one command-line argument present.

Since `--component` is not a required argument for out case.
This should be treat as 0 or more, not 1 or more.

Thanks.

peterzhuamazon commented on src/sign_workflow/sign_args.py:
```diff
@@ -0,0 +1,45 @@
+    def __init__(self) -> None:
+        parser = argparse.ArgumentParser(description="Sign artifacts")
+        parser.add_argument("target", type=Path, help="Path to local manifest file or artifact directory")
+        parser.add_argument("-c", "--component", type=str, nargs='+', dest="components", help="Component or components to sign")
```
Same as above.

peterzhuamazon commented on src/test_workflow/test_args.py:
```diff
@@ -28,14 +29,14 @@ def __init__(self):
         parser.add_argument("-p", "--paths", nargs='*', action=TestKwargs, default={}, help="Specify paths for OpenSearch and OpenSearch Dashboards.")
         parser.add_argument("--test-run-id", type=int, help="The unique execution id for the test")
-        parser.add_argument("--component", type=str, help="Test a specific component instead of the entire distribution.")
+        parser.add_argument("--component", type=str, dest="components", nargs='+', help="Test a specific component or components instead of the entire distribution.")
```
Same as above.

peterzhuamazon commented on src/ci_workflow/ci_args.py:
```diff
@@ -24,7 +26,14 @@ def __init__(self) -> None:
+            "--component",
+            type=str,
+            dest="components",
+            nargs='+',
```
Same as above.

peterzhuamazon reviewed (changes requested):
Thanks @dblock I do have some questions on several parts.
Thanks.

peterzhuamazon commented:
Also @dblock please update the jenkins file of both distribution builds params description, so user can take multiple components with spaces.
Thanks.

dblock commented on src/manifests/component_manifest.py:
```diff
@@ -51,19 +51,24 @@ def __to_dict__(self) -> List[Dict[Any, Any]]:
         """
+        if focus and len(focus) > 0:
+            invalid = [item for item in focus if item not in self]
+            if len(invalid) > 0:
+                raise ValueError(f"Unknown component{'s'[:len(invalid)^1]}={','.join(invalid)}.")
+
```
I think it's actually hard to read. Will change it to `== 1`, I got it from SO.

dblock commented on src/manifests/component_manifest.py:
```diff
@@ -79,8 +84,14 @@ def __to_dict__(self) -> dict:
-    def __matches__(self, focus: str = None, platform: str = None) -> bool:
-        matches = ((not focus) or (self.name == focus)) and ((not platform) or (not self.platforms) or (platform in self.platforms))
+    def __matches__(self, focus: List[str] = [], platform: str = None) -> bool:
+        matches = True
+
+        if focus and len(focus) > 0:
+            matches = matches and self.name in focus
+
+        if platform and self.platforms:
+            matches = matches and platform in self.platforms
```
Right. I just made this code more readable.

dblock commented:
> Also @dblock please update the jenkins file of both distribution builds params description, so user can take multiple components with spaces. Thanks.

Done. I didn't change `COMPONENT_NAME` to `COMPONENT_NAMES` or anything, but lmk if you think we should. I didn't want to break backwards compatibility in build.sh either and left it as `--component`.

peterzhuamazon commented on src/sign_workflow/sign_args.py:
```diff
@@ -0,0 +1,45 @@
+    def __init__(self) -> None:
+        parser = argparse.ArgumentParser(description="Sign artifacts")
+        parser.add_argument("target", type=Path, help="Path to local manifest file or artifact directory")
+        parser.add_argument("-c", "--component", type=str, nargs='+', dest="components", help="Component or components to sign")
```
@dblock this one still stay at +, not * please change.

dblock replied:
I haven't changed it.

peterzhuamazon replied:
I am a bit confused are you saying this one should stay on + or you are planning to change it later?
Thanks.

dblock replied:
Oh I thought it was a required parameter.

peterzhuamazon replied:
```
signArtifacts(
                            artifactPath: "$WORKSPACE/artifacts",
                            sigtype: SIGNATURE_TYPE,
                            platform: DISTRIBUTION_PLATFORM
                    )
```

For reference we dont need `--component` to be at least 1 to sign.

peterzhuamazon commented on jenkins/opensearch-dashboards/distribution-build.jenkinsfile:
```diff
@@ -14,7 +14,7 @@ pipeline {
         string(
             name: 'COMPONENT_NAME',
-            description: 'If this field contains a component name (e.g. OpenSearch-Dashboards), will build with "--component <COMPONENT_NAME>", else build everything in the INPUT_MANIFEST.',
+            description: 'If this field contains one or more component names (e.g. OpenSearch), will build with "--component <COMPONENT_NAME> ...", else build everything in the INPUT_MANIFEST.',
```
Can we change the `e.g.` section to be `(e.g. OpenSearch-Dashboards reportsDashboards ...)`

peterzhuamazon commented on jenkins/opensearch/distribution-build.jenkinsfile:
```diff
@@ -12,7 +12,7 @@ pipeline {
         string(
             name: 'COMPONENT_NAME',
-            description: 'If this field contains a component name (e.g. OpenSearch), will build with "--component <COMPONENT_NAME>", else build everything in the INPUT_MANIFEST.',
+            description: 'If this field contains one or more component names (e.g. OpenSearch), will build with "--component <COMPONENT_NAME> ...", else build everything in the INPUT_MANIFEST.',
```
Can we change the e.g. section to be `(e.g. OpenSearch common-utils ...)`

peterzhuamazon reviewed (changes requested):
Some missing parts.

peterzhuamazon commented on src/manifests/component_manifest.py:
```diff
@@ -51,19 +51,24 @@ def __to_dict__(self) -> List[Dict[Any, Any]]:
+        if focus and len(focus) > 0:
+            invalid = [item for item in focus if item not in self]
+            if len(invalid) > 0:
+                raise ValueError(f"Unknown component{'s'[:len(invalid) != 1]}={','.join(invalid)}.")
```
Using `len(invalid)^1` will return 0 or 1.
Using `len(invalid) != 1` will return True or False.

Apparently `'s'[:True]` will resolve as `'s'[:1]` and print `s`.
Great thanks.

peterzhuamazon replied:
```
In Python 3.x True and False are keywords and will always be equal to 1 and 0.
```

Not in Python 2.x tho glad our code base is in 3.

peterzhuamazon commented:
Thanks @dblock for this quick PR will use this to build 2.0.0.
Thanks.
